mllib/src/main/scala/org/apache/spark/mllib/fpm/Prefixspan.scala

```diff
@@ -0,0 +1,183 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.fpm
+
+import org.apache.spark.rdd.RDD
+
+/**
+ *
+ * A parallel PrefixSpan algorithm to mine sequential pattern.
+ * The PrefixSpan algorithm is described in
+ * [[http://web.engr.illinois.edu/~hanj/pdf/span01.pdf]].
+ *
+ * @param sequences original sequences data
+ * @param minSupport the minimal support level of the sequential pattern, any pattern appears
+ *                   more than minSupport times will be output
+ * @param maxPatternLength the maximal length of the sequential pattern, any pattern appears
+ *                   less than maxPatternLength will be output
+ *
+ * @see [[https://en.wikipedia.org/wiki/Sequential_Pattern_Mining Sequential Pattern Mining
+ *       (Wikipedia)]]
+ */
+class Prefixspan(
+    val sequences: RDD[Array[Int]],
```

Comment:
The Int data type is small than some other data types, such as String etc. So, inside the algorithm, we use Int data type to get the better performance.

If user use other data type, like String, maybe he will code String to Int, and decode the result to String. The following:

    val sequences = Array(
      Array("3", "1", "3", "4", "5"),
      Array("2", "3", "1"),
      Array("3", "4", "4", "3"),
      Array("1", "3", "4", "5"),
      Array("2", "4", "1"),
      Array("6", "5", "3"))
    val rdd = sc.parallelize(sequences, 2).cache()

    // create coder and decoder
    val letterMap = rdd.flatMap(x => x.distinct).distinct().zipWithIndex().mapValues(_.toInt).collect
    val coder = letterMap.toMap
    val decoder = letterMap.map(x => (x._2, x._1)).toMap

	// code
    val intRdd = rdd.map(x => x.map(y => coder(y)))

    val prefixspan1 = new Prefixspan(intRdd, 2, 50)
    val result = prefixspan1.run()

    // decode
    val stringResult = result.map(x => (x._1.map(y => decoder(y)), x._2))

I think this is a general job, some other algorithms maybe need this function too, so can we add the code and decoder as a separate model for all other algorithms ?